relation/src/gadgets/ecc/emulated/short_weierstrass.rs

```diff
@@ -0,0 +1,406 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! Short Weierstrass curve point addition
+
+use crate::{
+    errors::CircuitError,
+    gadgets::{EmulatedVariable, EmulationConfig},
+    BoolVar, Circuit, PlonkCircuit,
+};
+use ark_ec::short_weierstrass::{Affine, SWCurveConfig};
+use ark_ff::PrimeField;
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use ark_std::{vec, vec::Vec};
+
+/// An elliptic curve point in short Weierstrass affine form (x, y, infinity).
+#[derive(Debug, Eq, PartialEq, Copy, Clone, CanonicalSerialize, CanonicalDeserialize)]
+pub struct SWPoint<F: PrimeField>(pub F, pub F, pub bool);
+
+impl<F, P> From<Affine<P>> for SWPoint<F>
+where
+    F: PrimeField,
+    P: SWCurveConfig<BaseField = F>,
+{
+    fn from(p: Affine<P>) -> Self {
+        SWPoint(p.x, p.y, p.infinity)
+    }
+}
+
+/// The variable represents an SW point in the emulated field.
+#[derive(Debug, Clone)]
+pub struct EmulatedSWPointVariable<E: PrimeField>(
+    pub EmulatedVariable<E>,
+    pub EmulatedVariable<E>,
+    pub BoolVar,
+);
+
+impl<F: PrimeField> PlonkCircuit<F> {
+    /// Return the witness point
+    pub fn emulated_sw_point_witness<E: EmulationConfig<F>>(
+        &self,
+        point_var: &EmulatedSWPointVariable<E>,
+    ) -> Result<SWPoint<E>, CircuitError> {
+        let x = self.emulated_witness(&point_var.0)?;
+        let y = self.emulated_witness(&point_var.1)?;
+        let infinity = self.witness(point_var.2 .0)? == F::one();
+        Ok(SWPoint(x, y, infinity))
+    }
+
+    /// Add a new emulated EC point (as witness)
+    pub fn create_emulated_sw_point_variable<E: EmulationConfig<F>>(
+        &mut self,
+        point: SWPoint<E>,
+    ) -> Result<EmulatedSWPointVariable<E>, CircuitError> {
+        let x = self.create_emulated_variable(point.0)?;
+        let y = self.create_emulated_variable(point.1)?;
+        let infinity = self.create_boolean_variable(point.2)?;
+        Ok(EmulatedSWPointVariable(x, y, infinity))
+    }
+
+    /// Add a new constant emulated EC point
+    pub fn create_constant_emulated_sw_point_variable<E: EmulationConfig<F>>(
+        &mut self,
+        point: SWPoint<E>,
+    ) -> Result<EmulatedSWPointVariable<E>, CircuitError> {
+        let x = self.create_constant_emulated_variable(point.0)?;
+        let y = self.create_constant_emulated_variable(point.1)?;
+        let infinity =
+            self.create_boolean_variable_unchecked(if point.2 { F::one() } else { F::zero() })?;
```

Comment:
Since it's a constant, instead of creating a bool variable, it's better to just set it to `self.zero()` or `self.one()`

Reply:
Good idea, done in [e058b8a](https://github.com/EspressoSystems/jellyfish/pull/326/commits/e058b8a0f24648fe37a76575a87dbbbad21726dd)